diff --git a/node_modules/react-native-gesture-handler/android/build.gradle b/node_modules/react-native-gesture-handler/android/build.gradle
index 88c1bfc..3e39a52 100644
--- a/node_modules/react-native-gesture-handler/android/build.gradle
+++ b/node_modules/react-native-gesture-handler/android/build.gradle
@@ -1,4 +1,5 @@
 apply plugin: 'com.android.library'
+apply from: './expo/linking.gradle'
 
 def safeExtGet(prop, fallback) {
     rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
diff --git a/node_modules/react-native-gesture-handler/android/expo/linking.gradle b/node_modules/react-native-gesture-handler/android/expo/linking.gradle
new file mode 100644
index 0000000..512b2d2
--- /dev/null
+++ b/node_modules/react-native-gesture-handler/android/expo/linking.gradle
@@ -0,0 +1,20 @@
+def isExpoLinked = false
+rootProject.getSubprojects().forEach({ project ->
+    if (project.plugins.hasPlugin("com.android.application")) {
+        if (project.configurations.implementation.getDependencies().find { dep -> dep.name == "expo" }) {
+            isExpoLinked = true
+        }
+    }
+})
+
+if (isExpoLinked) {
+    android {
+        sourceSets {
+            main.java.srcDirs += 'expo/src/main/java'
+        }
+    }
+
+    dependencies {
+        implementation project(':expo-modules-core')
+    }
+}
diff --git a/node_modules/react-native-gesture-handler/android/expo/src/main/java/expo/modules/adapters/gesturehandler/EXGestureHandlerPackage.java b/node_modules/react-native-gesture-handler/android/expo/src/main/java/expo/modules/adapters/gesturehandler/EXGestureHandlerPackage.java
new file mode 100644
index 0000000..b42b48e
--- /dev/null
+++ b/node_modules/react-native-gesture-handler/android/expo/src/main/java/expo/modules/adapters/gesturehandler/EXGestureHandlerPackage.java
@@ -0,0 +1,28 @@
+package expo.modules.adapters.gesturehandler;
+
+import android.app.Activity;
+import android.content.Context;
+import android.os.Bundle;
+
+import com.facebook.react.ReactRootView;
+import com.swmansion.gesturehandler.react.RNGestureHandlerEnabledRootView;
+
+import java.util.Collections;
+import java.util.List;
+
+import expo.modules.core.interfaces.Package;
+import expo.modules.core.interfaces.ReactActivityHandler;
+
+public class EXGestureHandlerPackage implements Package {
+    @Override
+    public List<ReactActivityHandler> createReactActivityHandlers(Context activityContext) {
+        final ReactActivityHandler handler = new ReactActivityHandler() {
+            @Override
+            public ReactRootView createReactRootView(Activity activity) {
+                return new RNGestureHandlerEnabledRootView(activity);
+            }
+        };
+
+        return Collections.singletonList(handler);
+    }
+}
diff --git a/node_modules/react-native-gesture-handler/expo-module.config.json b/node_modules/react-native-gesture-handler/expo-module.config.json
new file mode 100644
index 0000000..1666f66
--- /dev/null
+++ b/node_modules/react-native-gesture-handler/expo-module.config.json
@@ -0,0 +1,4 @@
+{
+  "name": "react-native-gesture-handler",
+  "platforms": ["android"]
+}
