diff --git a/node_modules/react-native-reanimated/android/build.gradle b/node_modules/react-native-reanimated/android/build.gradle
index 18ee54a..5ecd39f 100644
--- a/node_modules/react-native-reanimated/android/build.gradle
+++ b/node_modules/react-native-reanimated/android/build.gradle
@@ -83,3 +83,5 @@ rootProject.getSubprojects().forEach({project ->
 })
 
 artifacts.add("default", file("react-native-reanimated-${minor}-${engine}.aar"))
+
+apply from: './expo/linking.gradle'
diff --git a/node_modules/react-native-reanimated/android/expo/linking.gradle b/node_modules/react-native-reanimated/android/expo/linking.gradle
new file mode 100644
index 0000000..3a583a6
--- /dev/null
+++ b/node_modules/react-native-reanimated/android/expo/linking.gradle
@@ -0,0 +1,34 @@
+def isExpoLinked = false
+rootProject.getSubprojects().forEach({ project ->
+    if (project.plugins.hasPlugin("com.android.application")) {
+        if (project.configurations.implementation.getDependencies().find { dep -> dep.name == "expo" }) {
+            isExpoLinked = true
+        }
+    }
+})
+
+def safeExtGet(prop, fallback) {
+    rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
+}
+
+if (isExpoLinked) {
+    apply plugin: 'com.android.library'
+
+    android {
+        compileSdkVersion safeExtGet('compileSdkVersion', 30)
+
+        sourceSets {
+            main.manifest.srcFile 'src/AndroidManifest.xml'
+            main.java.srcDirs = [ 'expo/src/main/java' ]
+        }
+    }
+
+    dependencies {
+        implementation 'com.facebook.react:react-native:+'
+        implementation project(':expo-modules-core')
+
+        project.configurations.default.artifacts.each { artifact ->
+            api files(artifact.file)
+        }
+    }
+}
diff --git a/node_modules/react-native-reanimated/android/expo/src/main/java/com/swmansion/reanimated/EXReanimatedAdapter.java b/node_modules/react-native-reanimated/android/expo/src/main/java/com/swmansion/reanimated/EXReanimatedAdapter.java
new file mode 100644
index 0000000..da76cd4
--- /dev/null
+++ b/node_modules/react-native-reanimated/android/expo/src/main/java/com/swmansion/reanimated/EXReanimatedAdapter.java
@@ -0,0 +1,12 @@
+package com.swmansion.reanimated;
+
+import com.facebook.react.bridge.JavaScriptContextHolder;
+import com.facebook.react.bridge.ReactApplicationContext;
+
+public class EXReanimatedAdapter {
+  public static void registerJSIModules(ReactApplicationContext reactApplicationContext, JavaScriptContextHolder jsContext) {
+    NodesManager nodesManager =
+      reactApplicationContext.getNativeModule(ReanimatedModule.class).getNodesManager();
+    nodesManager.initWithContext(reactApplicationContext);
+  }
+}
diff --git a/node_modules/react-native-reanimated/android/expo/src/main/java/expo/modules/adapters/reanimated/EXReanimatedPackage.java b/node_modules/react-native-reanimated/android/expo/src/main/java/expo/modules/adapters/reanimated/EXReanimatedPackage.java
new file mode 100644
index 0000000..199e9fd
--- /dev/null
+++ b/node_modules/react-native-reanimated/android/expo/src/main/java/expo/modules/adapters/reanimated/EXReanimatedPackage.java
@@ -0,0 +1,52 @@
+package expo.modules.adapters.reanimated;
+
+import android.content.Context;
+
+import com.facebook.react.ReactInstanceManager;
+import com.facebook.react.bridge.JavaScriptContextHolder;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.swmansion.reanimated.EXReanimatedAdapter;
+
+import java.util.Collections;
+import java.util.List;
+
+import expo.modules.core.interfaces.Package;
+import expo.modules.core.interfaces.ReactNativeHostHandler;
+
+public class EXReanimatedPackage implements Package {
+  @Override
+  public List<? extends ReactNativeHostHandler> createReactNativeHostHandlers(Context context) {
+    final ReactNativeHostHandler handler = new ReactNativeHostHandler() {
+      @Override
+      public ReactInstanceManager createReactInstanceManager(boolean useDeveloperSupport) {
+        return null;
+      }
+
+      @Override
+      public String getJSBundleFile(boolean useDeveloperSupport) {
+        return null;
+      }
+
+      @Override
+      public String getBundleAssetName(boolean useDeveloperSupport) {
+        return null;
+      }
+
+      @Override
+      public void onRegisterJSIModules(ReactApplicationContext reactApplicationContext, JavaScriptContextHolder jsContext, boolean useDeveloperSupport) {
+        EXReanimatedAdapter.registerJSIModules(reactApplicationContext, jsContext);
+      }
+
+      @Override
+      public void onWillCreateReactInstanceManager(boolean useDeveloperSupport) {
+
+      }
+
+      @Override
+      public void onDidCreateReactInstanceManager(boolean useDeveloperSupport) {
+
+      }
+    };
+    return Collections.singletonList(handler);
+  }
+}
diff --git a/node_modules/react-native-reanimated/expo-module.config.json b/node_modules/react-native-reanimated/expo-module.config.json
new file mode 100644
index 0000000..7bf118d
--- /dev/null
+++ b/node_modules/react-native-reanimated/expo-module.config.json
@@ -0,0 +1,4 @@
+{
+  "name": "react-native-reanimated",
+  "platforms": ["android"]
+}
